pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/AnnotationSuppressionUtil.java

```diff
@@ -0,0 +1,115 @@
+ */
+
+package net.sourceforge.pmd.lang.java.rule;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import net.sourceforge.pmd.Rule;
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.lang.java.ast.ASTAnnotation;
+import net.sourceforge.pmd.lang.java.ast.ASTAnyTypeDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTFieldDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTFormalParameter;
+import net.sourceforge.pmd.lang.java.ast.ASTLiteral;
+import net.sourceforge.pmd.lang.java.ast.ASTLocalVariableDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTMethodOrConstructorDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTTypeDeclaration;
+import net.sourceforge.pmd.lang.java.ast.Annotatable;
+import net.sourceforge.pmd.lang.java.typeresolution.TypeHelper;
+
+/**
+ * Helper methods to suppress violations based on annotations.
+ *
+ * An annotation suppresses a rule if the annotation is a {@link SuppressWarnings},
+ * and if the set of suppressed warnings ({@link SuppressWarnings#value()})
+ * contains at least one of those:
+ * <ul>
+ * <li>"PMD" (suppresses all rules);
+ * <li>"PMD.rulename", where rulename is the name of the given rule;
+ * <li>"all" (conventional value to suppress all warnings).
+ * </ul>
+ *
+ * <p>Additionally, the following values suppress a specific set of rules:
+ * <ul>
+ * <li>{@code "unused"}: suppresses rules like UnusedLocalVariable or UnusedPrivateField;
+ * <li>{@code "serial"}: suppresses BeanMembersShouldSerialize and MissingSerialVersionUID;
+ * <li>TODO "fallthrough" #1899
+ * </ul>
+ */
+final class AnnotationSuppressionUtil {
+
+    private static final Set<String> UNUSED_RULES
+        = new HashSet<>(Arrays.asList("UnusedPrivateField", "UnusedLocalVariable", "UnusedPrivateMethod", "UnusedFormalParameter"));
+    private static final Set<String> SERIAL_RULES =
+        new HashSet<>(Arrays.asList("BeanMembersShouldSerialize", "MissingSerialVersionUID"));
+
+    private AnnotationSuppressionUtil() {
+
+    }
+
+    /**
+     * Returns true if the node has an annotation that suppresses the
+     * given rule.
+     */
+    static boolean suppresses(final Node node, Rule rule) {
+        Annotatable suppressor = getSuppressor(node);
+        if (suppressor == null) {
+            return false;
+        }
+
+        return hasSuppressWarningsAnnotationFor(suppressor, rule);
+    }
+
+    @Nullable
+    private static Annotatable getSuppressor(Node node) {
+        if (node instanceof ASTAnyTypeDeclaration
+            || node instanceof ASTMethodOrConstructorDeclaration
+            // also works for ASTResource when Resource uses LocalVariableDeclaration
+            || node instanceof ASTLocalVariableDeclaration
+            || node instanceof ASTFieldDeclaration
+            || node instanceof ASTFormalParameter) {
```

Comment:
should we simplify this check to simply use: `if (node instanceof Annotatable) {`?

It would be more concise, and would not assume that any of these types are `Annotatable` on their own, which should be an implementation detail. If a new node is introduced and implements the annotation it would work out of the box.

Reply:
SuppressWarnings has the following targeting policy:
```java
@Target({TYPE, FIELD, METHOD, PARAMETER, CONSTRUCTOR, LOCAL_VARIABLE, MODULE})
@Retention(RetentionPolicy.SOURCE)
public @interface SuppressWarnings {
```
Most noticeably, it cannot be used as a type annotation (`TYPE_USE`). But many nodes implement Annotatable while only supporting type annotations. So I think this method should just target what SuppressWarnings can do (I just noticed, we're missing modules too), and I'm not sure what being that general would serve here.

Besides, I don't feel like those types' implementing Annotatable is an implementation detail.. it's part of our published API